Add onRender hook

shishu314 opened the pull request:
**Problem**
---
We are trying to see how long frames take to render in Gemini CLI and would like to make changes upstream in ink to allow access to this information. 

**Changes**
---
Add an option to listen for rendering time, this should not break any existing code.

**Testing Plans**
---
1. Run the test to make sure that the callback function is called on either `render` or `re-render`.

sindresorhus commented:
Your `onRender` callback runs only for manual `render()`/`rerender()` calls. Internal React updates call `rootNode.onRender` → `Ink.onRender()`, which never invokes `options.onRender`.

sindresorhus commented on src/render.ts:
```diff
@@ -42,6 +42,11 @@ export type RenderOptions = {
+	/**
+	Function to call after each render.
+	*/
+	onRender?: (renderTime: number) => void;
```
I think it should be an object with renderTime property in case we want to add more things in the future.

sindresorhus commented on src/ink.tsx:
```diff
@@ -280,12 +279,17 @@ export default class Ink {
 			</AccessibilityContext.Provider>
 		);
 
+		const start = Date.now();
```
Use `performance.now();`

shishu314 commented:
> Your `onRender` callback runs only for manual `render()`/`rerender()` calls. Internal React updates call `rootNode.onRender` → `Ink.onRender()`, which never invokes `options.onRender`.

Addressed this and added the callback into `Ink.onRender()`

shishu314 commented on test/render.tsx:
```diff
@@ -273,3 +273,48 @@ test.serial('throttle renders to maxFps', t => {
 	}
 });
+
+test.serial('outputs renderTime when onRender is passed', t => {
```
I added a case to check for internal state updates when stdin emits. However it seems like `onRender` only triggers once and does not advance the clock (see assertion). I am not sure why this is the case (not too familiar with react lifecycles but it seems like it should re-render). Any pointers on why this is the case?
